pkg/collector/corechecks/servicediscovery/discover_services_linux.go

```diff
@@ -0,0 +1,318 @@
+}
+
+type aliveMap struct {
+	m  map[int]*processInfo
+	mu sync.RWMutex
+}
+
+func (a *aliveMap) get(pid int) (*processInfo, bool) {
+	a.mu.RLock()
+	defer a.mu.RUnlock()
+	v, ok := a.m[pid]
+	return v, ok
+}
+
+func (a *aliveMap) set(pid int, info *processInfo) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.m[pid] = info
+}
+
+func (a *aliveMap) delete(pid int) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	delete(a.m, pid)
+}
+
+type ignoreMap struct {
+	m  map[int]string
+	mu sync.RWMutex
+}
+
+func (i *ignoreMap) get(pid int) (string, bool) {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	v, ok := i.m[pid]
+	return v, ok
+}
+
+func (i *ignoreMap) set(pid int, name string) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.m[pid] = name
+}
+
+func (i *ignoreMap) delete(pid int) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	delete(i.m, pid)
+}
+
+type linuxImpl struct {
+	sender    *telemetrySender
+	ignoreCfg map[string]struct{}
+
+	// PID -> process info
+	// pids that are considered services will be added here.
+	alive *aliveMap
+
+	// PID -> process name
+	// pids will be added here because the name was ignored in the config, or because
+	// it was already scanned and had no open ports.
+	ignore *ignoreMap
+}
+
+func newLinuxImpl(sender *telemetrySender, ignoreProcesses map[string]struct{}) osImpl {
+	return &linuxImpl{
+		sender:    sender,
+		ignoreCfg: ignoreProcesses,
+		ignore: &ignoreMap{
+			m: make(map[int]string),
+		},
+		alive: &aliveMap{
+			m: make(map[int]*processInfo),
+		},
+	}
+}
+
+func (i *linuxImpl) DiscoverServices() error {
+	processes, err := procfs.AllProcs()
+	if err != nil {
+		return err
+	}
+
+	portPoller, err := portlist.NewPoller(false)
+	if err != nil {
+		return fmt.Errorf("failed to initialize port poller: %w", err)
+	}
+	defer func() {
+		if err := portPoller.Close(); err != nil {
+			log.Warnf("failed to close port poller: %v", err)
+		}
+	}()
+
+	openPorts, err := portPoller.ListeningPorts()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("alive: %d | services: %d | ignore: %d", len(processes), len(i.alive.m), len(i.ignore.m))
+	aliveProcesses := make(map[int]struct{})
+
+	var wg sync.WaitGroup
+
+	for _, p := range processes {
+		aliveProcesses[p.PID] = struct{}{}
+		name, shortName, err := processName(p)
+		if err != nil {
+			log.Errorf("[pid: %d] failed to get process name: %v", p.PID, err)
+			continue
+		}
+
+		ignore := false
+		if ignoreName, ok := i.ignore.get(p.PID); ok {
+			// ignore
+			if ignoreName == name {
+				ignore = true
+			} else {
+				// same pid but different process name, this means the pid has been reused for
+				// a different process
+				i.ignore.delete(p.PID)
+			}
+		} else if _, ok := i.ignoreCfg[shortName]; ok {
+			ignore = true
+			i.ignore.set(p.PID, name)
+		}
+
+		if !ignore {
+			// process is a potential service
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				i.scanProcess(p.PID, openPorts)
+			}()
+		}
+	}
+
+	for pid, pInfo := range i.alive.m {
+		_, ok := aliveProcesses[pid]
```

Comment:
This is the reason why you should use map[T]bool for a set. The test to see if something matches becomes just `if aliveProcesses[pid]`